### ISSUE TITLE
Ambiguous operator == leads to nil comparisons being serialized wrongly

### ISSUE THREAD

**Issue by MrMage** [vapor/sql#13]
The operator `==` is defined twice, in ambiguous ways:

https://github.com/vapor/fluent/blob/f8378e392ed6d2b2f55c88d6c59ede96e74897fa/Sources/Fluent/QueryBuilder/QueryBuilder%2BOperators.swift#L36

https://github.com/vapor/sql/blob/9c8c19bd210b927ee52532d204e255efb11508a4/Sources/SQL/SQLBinaryOperator.swift#L180

The problem is that the second one does not serialize `== nil` comparisons as `IS NULL`.

The `Model` implementation of `withSoftDeleted` (see https://github.com/vapor/fluent/blob/f8378e392ed6d2b2f55c88d6c59ede96e74897fa/Sources/Fluent/Model/Model.swift#L379) uses the correct operator (the first one), while my own `and.filter(\RefreshToken.deletedAt == nil)` uses the second operator, which leads to something like `"deletedAt" = NULL` (or `= <some object>`) being encoded instead of `"deletedAt" IS NULL`.

**Comment by proggeramlug** [comment]
@MrMage @tanner0101 

I driled down the bug to BinaryOperators and corrected it by adding a condition to check for nil. SQL queries look as expected now.
PR: https://github.com/vapor/sql/pull/14

**Comment by MrMage** [comment]
I feel like the underlying issue is that there are _two_ different operators and it's unclear which one gets chosen in a particular solution. So I think the proper solution would be two merge these two operators to eliminate any ambiguity. @tanner0101, could you provide insight into the difference between these two operators at the moment?

**Comment by proggeramlug** [comment]
That's what I observed as well. I think the reason the operators of `QueryBuilder` are not fired here is because `nil` is not of type `Value` and so the BinaryOperator takes that one.
I'm not sure why would not combine the two though, it is confusing.

**Comment by tanner0101** [comment]
@MrMage the problem here is [this overload](https://github.com/vapor/fluent/blob/master/Sources/Fluent/QueryBuilder/QueryBuilder%2BFilter.swift#L91). That accepts the DB's generic filter type, in the case of MySQL, that's `MySQLExpression`.

vapor/mysql has it's own, lower level query builder based on vapor/sql. This query builder is similar to Fluent in some ways, for example you can do:

```swift
where(\Model.prop == foo)
```

More examples [here](https://github.com/vapor/sql/blob/master/Sources/SQLBenchmark/SQLBenchmark%2BTestPlanets.swift#L69-L71).

This is of course based on operators that generate `MySQLExpression`. Because Fluent has an overload that accepts that, the ambiguity is happening.

My solution as recommended in https://github.com/vapor/fluent-mysql/issues/120 is to either remove that overload altogether or give it a prefix like `.filter(raw: ...)`.

**Comment by MrMage** [comment]
I agree that these two should be different to avoid any possible ambiguity. It's bad user experience that a statement might work differently depending on whether `FluentPostgreSQL`/`FluentMySQL` is imported or not. Also, **both methods should be documented to make it clear which one is the recommended one (and how their behavior might differ).**

**Comment by tanner0101** [comment]
Sorry, linked the wrong commit. Here it is: https://github.com/vapor/fluent/commit/c5c384a9af63a3c55e7ff164c39a5c037c6443d1#diff-0b1072a5a6a4f0127933ea67791ec0d8
